### ISSUE TITLE
Filtering by binary field doesn't work on Oracle

### ISSUE THREAD

**Issue by morozov** [doctrine/dbal#2787]
Setup:
```php
$table = new \Doctrine\DBAL\Schema\Table('test');
$table->addColumn('id', 'integer');
$table->addColumn('value', 'binary', ['length' => 16]);

$sm = $conn->getSchemaManager();
$sm->createTable($table);

$conn->insert('test', [
    'id' => 1,
    'value' => 'test',
], [
    PDO::PARAM_INT,
    PDO::PARAM_LOB,
]);
```
Binding param as LOB (same as during insertion):
```php
$stmt = $conn->executeQuery('SELECT id FROM test WHERE value = ?', ['test'], [PDO::PARAM_LOB]);
var_dump($stmt->fetch());

// ORA-00932: inconsistent datatypes: expected - got BLOB
```

Binding param as a string:
```php
$stmt = $conn->executeQuery('SELECT id FROM test WHERE value = ?', ['test']);
var_dump($stmt->fetch());

// bool(false) – not found
```

Representing the value as a hexadecimal string:
```php
$stmt = $conn->executeQuery('SELECT id FROM test WHERE value = ?', [bin2hex('test')]);
var_dump($stmt->fetch());

// array(1) {
//  'ID' =>
//  string(1) "1"
// }
```

Binding param as a binary string:
```php
$stmt = oci_parse($conn, 'SELECT id FROM test WHERE value = :param');
$param = 'test';
oci_bind_by_name($stmt, 'param', $param, -1, OCI_B_BIN);
oci_execute($stmt);
var_dump(oci_fetch_array($stmt, OCI_ASSOC));

// array(1) {
//  'ID' =>
//  string(1) "1"
// }
```

**Comment by Ocramius** [comment]
We can't really detect the server-side parameter type on our own, but I suppose that we could add `bin2hex()` when `PDO::PARAM_LOB` is used.

What happens for insert operations?

**Comment by morozov** [comment]
> What happens for insert operations?

I think the fact that binding as LOB for INSERT/UPDATE happens to work is an unintended behavior on the server/driver side. Probably, during a write operation, Oracle doesn't really care if it's a string or a stream — it just gets written, but during the comparison, it cannot compare anything with a LOB by design.

If there was a PDO constant for the binary type, we could use it for both reading and writing and not rely on this behavior. I tried to find a reference of how binary values can be bound in PDO (something like [`PDO::SQLSRV_ENCODING_BINARY`](https://github.com/doctrine/dbal/blob/v2.6.0/lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php#L36)) but didn't manage to, and I wouldn't be surprised if it's not supported at all.

I agree, the approach with `bin2hex()` looks the most realistic in our case. Do we need to introduce some new `Platform` method which would call `bin2hex()` for Oracle?

**Comment by Ocramius** [comment]
> I agree, the approach with `bin2hex()` looks the most realistic in our case. Do we need to introduce some new `Platform` method which would call `bin2hex()` for Oracle?

@morozov yeah, something to process LOB before passing it in, I suppose. I don't know if we are already processing bound parameters at all though (besides type conversions)

**Comment by morozov** [comment]
Fixed by #3149.
